include/rocksdb/env_encrypt2.h

```diff
@@ -0,0 +1,471 @@
+  //         NotFound if "dir" does not exist, the calling process does not have
+  //                  permission to access "dir", or if "dir" is invalid.
+  //         IOError if an IO Error was encountered
+  virtual Status GetChildrenFileAttributes(
+      const std::string& dir, std::vector<FileAttributes>* result) override;
+
+  // Store the size of fname in *file_size.
+  virtual Status GetFileSize(const std::string& fname,
+                             uint64_t* file_size) override;
+
+  // only needed for GetChildrenFileAttributes & GetFileSize
+  virtual Status GetEncryptionProvider(
+      const std::string& fname, std::shared_ptr<EncryptionProvider>& provider);
+
+  template <class TypeFile>
+  Status ReadSeqEncryptionPrefix(
+      TypeFile* f, std::shared_ptr<EncryptionProvider>& provider,
+      std::unique_ptr<BlockAccessCipherStream>& stream) {
+    Status status;
+
+    provider.reset();  // nullptr for provider implies "no encryption"
+    stream.release();
+
+    // Look for encryption marker
+    EncryptMarker_t marker;
+    Slice marker_slice;
+    status = f->Read(sizeof(marker), &marker_slice, marker);
+    if (status.ok()) {
+      if (sizeof(marker) == marker_slice.size() &&
+          marker_slice.starts_with(Marker)) {
+        // code_version currently unused
+        uint8_t code_version = (uint8_t)marker_slice[7];
+
+        Slice prefix_slice;
+        Prefix0_t prefix_buffer;
+        status =
+            f->Read(sizeof(Prefix0_t), &prefix_slice, (char*)&prefix_buffer);
+        if (status.ok() && sizeof(Prefix0_t) == prefix_slice.size()) {
+          Sha1Description_t desc(prefix_buffer.key_description_,
+                                 sizeof(prefix_buffer.key_description_));
+
+          auto it = encrypt_read_.find(desc);
+          if (encrypt_read_.end() != it) {
+            CTREncryptionProvider2* ptr =
+                (CTREncryptionProvider2*)it->second.get();
+            provider = it->second;
+            stream.reset(new AESBlockAccessCipherStream(
+                ptr->key(), code_version, prefix_buffer.nonce_));
+          } else {
+            status = Status::NotSupported(
+                "No encryption key found to match input file");
+          }
+        }
+      }
+    }
+    return status;
+  }
+
+  template <class TypeFile>
+  Status ReadRandEncryptionPrefix(
+      TypeFile* f, std::shared_ptr<EncryptionProvider>& provider,
+      std::unique_ptr<BlockAccessCipherStream>& stream) {
+    Status status;
+
+    provider.reset();  // nullptr for provider implies "no encryption"
+    stream.release();
+
+    // Look for encryption marker
+    EncryptMarker_t marker;
+    Slice marker_slice;
+    status = f->Read(0, sizeof(marker), &marker_slice, marker);
+    if (status.ok()) {
+      if (sizeof(marker) == marker_slice.size() &&
+          marker_slice.starts_with(Marker)) {
+        // code_version currently unused
+        uint8_t code_version = (uint8_t)marker_slice[7];
```

Comment:
can we verify `code_version` is indeed 0, or return error?

Reply:
done

Reply:
Same for `ReadSeqEncryptionPrefix`.

Reply:
oops ... will fix.